src/data/setoid.lean

```diff
@@ -0,0 +1,603 @@
+  le_sup_left := λ r s, le_Inf _ r $ λ _ hx, hx.1,
+  le_sup_right := λ r s, le_Inf _ s $ λ _ hx, hx.2,
+  sup_le := λ r s t h1 h2, Inf_le _ t ⟨h1, h2⟩,
+  inf := has_inf.inf,
+  inf_le_left := λ _ _ _ _ h, h.1,
+  inf_le_right := λ _ _ _ _ h, h.2,
+  le_inf := λ _ _ _ h1 h2 _ _ h, ⟨h1 _ _ h, h2 _ _ h⟩,
+  top := ⟨λ _ _, true, ⟨λ _, trivial, λ _ _ h, h, λ _ _ _ h1 h2, h1⟩⟩,
+  le_top := λ _ _ _ _, trivial,
+  bot := ⟨(=), ⟨λ _, rfl, λ _ _ h, h.symm, λ _ _ _ h1 h2, h1.trans h2⟩⟩,
+  bot_le := λ r x y h, h ▸ r.2.1 x,
+  Sup := λ tt, Inf {t | ∀ t'∈tt, t' ≤ t},
+  Inf := has_Inf.Inf,
+  le_Sup := λ _ _ hs, le_Inf _ _ $ λ r hr, hr _ hs,
+  Sup_le := λ _ _ hs, Inf_le _ _ hs,
+  Inf_le := Inf_le,
+  le_Inf := le_Inf }
+
+/-- The inductively defined equivalence closure of a binary relation r is the infimum 
+    of the set of all equivalence relations containing r. -/
+theorem eqv_gen_eq (r : α → α → Prop) : 
+  eqv_gen.setoid r = Inf {s : setoid α | ∀ x y, r x y → s.rel x y} :=
+setoid.ext' $ λ _ _, 
+  ⟨λ H, eqv_gen.rec (λ _ _ h _ hs, hs _ _ h) (refl' _) 
+    (λ _ _ _ h, symm' _ h) (λ _ _ _ _ _ h1 h2, trans' _ h1 h2) H, 
+  Inf_le _ _ (λ _ _, eqv_gen.rel _ _) _ _⟩
+
+/-- The supremum of two equivalence relations r and s is the equivalence closure of the binary
+    relation 'x is related to y by r or s'. -/
+lemma sup_eq_eqv_gen (r s : setoid α) : 
+  r ⊔ s = eqv_gen.setoid (λ x y, r.rel x y ∨ s.rel x y) :=
+by rw eqv_gen_eq; apply congr_arg Inf; ext; exact 
+  ⟨λ h _ _ H, or.elim H (λ hl, h.1 _ _ hl) $ λ hr, h.2 _ _ hr, 
+   λ H, ⟨λ _ _ h, H _ _ $ or.inl h, λ _ _ h, H _ _ $ or.inr h⟩⟩
+
+/-- The supremum of a set S of equivalence relations is the equivalence closure of the binary
+    relation 'there exists r ∈ S relating x and y'. -/
+lemma Sup_eq_eqv_gen (S : set (setoid α)) : 
+  Sup S = eqv_gen.setoid (λ x y, ∃ r : setoid α, r ∈ S ∧ r.rel x y) :=
+by rw eqv_gen_eq; apply congr_arg Inf; ext;
+   exact ⟨λ h _ _ ⟨r, hr⟩, h r hr.1 _ _ hr.2, λ h r hS _ _ hr, h _ _ ⟨r, hS, hr⟩⟩
+
+/-- The equivalence closure of an equivalence relation r is r. -/
+@[simp] lemma eqv_gen_of_setoid (r : setoid α) : eqv_gen.setoid r.r = r :=
+le_antisymm (by rw eqv_gen_eq; exact Inf_le _ r (λ _ _ h, h)) eqv_gen.rel
+
+/-- Equivalence closure is idempotent. -/
+@[simp] lemma eqv_gen_idem (r : α → α → Prop) : 
+  eqv_gen.setoid (eqv_gen.setoid r).rel = eqv_gen.setoid r :=
+eqv_gen_of_setoid _ 
+
+/-- The equivalence closure of a binary relation r is contained in any equivalence
+    relation containing r. -/
+theorem eqv_gen_le {r : α → α → Prop} {s : setoid α} (h : ∀ x y, r x y → s.rel x y) : 
+  eqv_gen.setoid r ≤ s :=
+by rw eqv_gen_eq; exact Inf_le _ _ h
+
+/-- Equivalence closure of binary relations is monotonic. -/
+theorem eqv_gen_mono {r s : α → α → Prop} (h : ∀ x y, r x y → s x y) : 
+  eqv_gen.setoid r ≤ eqv_gen.setoid s :=
+eqv_gen_le $ λ _ _ hr, eqv_gen.rel _ _ $ h _ _ hr
+
+theorem le_def {r s : setoid α} : r ≤ s ↔ ∀ {x y}, r.rel x y → s.rel x y := iff.rfl
```

Comment:
Please move these `_def` theorems to directly after the definition of `le` (resp. `inf`, etc).